StorageProviders/AwsS3StorageProvider.cs

```diff
@@ -0,0 +1,91 @@
+using Amazon;
+using Amazon.S3;
+using Amazon.S3.Model;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Threading.Tasks;
+
+namespace StorageProviders
+{
+    public class AwsS3StorageProvider : IStorageProvider
+    {
+        private readonly AwsS3Settings _awsS3Settings;
+        private readonly ServiceSettings _serviceSettings;
+        private readonly ILogger<AwsS3StorageProvider> _logger;
+        private readonly Dictionary<string, RegionEndpoint> _regionEndpoints = new Dictionary<string, RegionEndpoint>
+        {
+            {"USEast1",  RegionEndpoint.USEast1},
+            {"USWest1",  RegionEndpoint.USWest1},
+            {"USWest2",  RegionEndpoint.USWest2},
+            {"EUWest1",  RegionEndpoint.EUWest1},
+            {"EUCentral1",  RegionEndpoint.EUCentral1},
+            {"APNortheast1",  RegionEndpoint.APNortheast1},
+            {"APSoutheast1",  RegionEndpoint.APSoutheast1},
+            {"APSoutheast2",  RegionEndpoint.APSoutheast2},
+            {"SAEast1",  RegionEndpoint.SAEast1},
+            {"USGovCloudWest1",  RegionEndpoint.USGovCloudWest1},
+            {"CNNorth1",  RegionEndpoint.CNNorth1}
+        };
+
+        public AwsS3StorageProvider(IOptions<AwsS3Settings> awsS3settings, IOptions<ServiceSettings> serviceSettings, ILogger<AwsS3StorageProvider> logger)
+        {
+            _awsS3Settings = awsS3settings.Value;
+            _serviceSettings = serviceSettings.Value;
+            _logger = logger;
+        }
+
+        protected virtual IAmazonS3 GetClient()
+        {
+            return new AmazonS3Client(_awsS3Settings.AccountInfo.AccessKey, _awsS3Settings.AccountInfo.SecretAccessKey, _regionEndpoints[_awsS3Settings.AccountInfo.RegionEndpoint]);
+        }
+
+        public async Task<UploadFileResult> UploadFile(string pathFile, string idUser, byte[] content, bool overwrite = false)
+        {
+            var request = new PutObjectRequest
+            {
+                InputStream = new MemoryStream(content),
+                BucketName = _awsS3Settings.BucketName,
+                Key = $"/{_serviceSettings.UsersRootFolder}/{idUser}{pathFile}"
```

Comment:
In my opinion, the key should not have a relation with a user-entered path.

It will give us problems.

Reply:
Aws S3 doesn't understand the folder and files. It uses an object key. If we check the path and file pattern we won't have a problem. I think using a path file as the key will allow us to use S3 as a file system.